contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,443 @@
+
+In downgrade scenario where the migration is then turned off we will have to
+remove these floating VolumeAttachment objects, the same issue is outlined above
+in the Non-Dynamic Provisioned Volumes section.
+
+
+## Interactions with PV-PVC Protection Finalizers
+PV-PVC Protection finalizers prevent deletion of a PV when it is bound to a PVC,
+and prevent deletion of a PVC when it is in use by a pod.
+
+There is no known issue with interaction here. The finalizers will still work in
+the same ways as we are not removing/adding PV’s or PVC’s in out of the ordinary
+ways.
+
+## Dealing with CSI Driver Failures
+Plugin should fail if the CSI Driver is down and migration is turned on. When
+the driver recovers we should be able to resume gracefully.
+
+We will also create a playbook entry for how to turn off the CSI Driver
+migration gracefully, how to tell when the CSI Driver is broken or non-existent,
+and how to redeploy a CSI Driver in a cluster.
+
+
+## Upgrade/Downgrade, Migrate/Un-migrate
+### Kubelet Node Annotation
+When the Kubelet starts, it will check whether the feature gate is
+enabled and if so will annotate its node with `csi.attach.kubernetes.io/gce-pd`
+for example to communicate to the A/D Controller that it supports migration of
+the gce-pd to CSI. The A/D Controller will have to choose on a per-node basis
+whether to use the CSI or the in-tree plugin for attach based on 3 criterea:
+1. Feature gate
+2. Plugin Migratable (Implements MigratablePlugin interface)
+3. Node to Attach to has requisite Annotation
+
+Note: All 3 criterea must be satisfied for A/D controller to Attach/Detach with
+CSI instead of in-tree plugin. For example if a Kubelet has feature on and marks
+the annotation, but the A/D Controller does not have the feature gate flipped,
+we consider this user error and will throw some errors.
+
+This can cause a race between the A/D Controller and the Kubelet annotating, if
+a volume is attached before the Kubelet completes annotation the A/D controller
+could attach using in-tree plugin instead of CSI while the Kubelet is expecting
+a CSI Attach. The same issue exists on downgrade if the Annotation is not
+removed before a volume is attached. An additional consideration is that we
+cannot have the Kubelet downgraded to a version that does not have the
+Annotation removal code.
+
+### Node Drain Requirement
+We require node's to be drained whenever the Kubelet is Upgrade/Downgraded or
+Migrated/Unmigrated to ensure that the entire volume lifecycle is maintained
+inside one code branch (CSI or In-tree). This simplifies upgrade/downgrade
+significantly and reduces chance of error and races.
+
+### Upgrade/Downgrade Migrate/Unmigrate Scenarios
+For upgrade, starting from a non-migrated cluster you must turn on migration for
+A/D Controller first, then drain your node before turning on migration for the
+Kubelet. The workflow is as follows:
+1. A/D Controller and Kubelet are both not migrated
```

Comment:
Adding "unschedulable" flag to kubelet to make sure that the node is annotated before it can accept first pods.
1. A/D Controller and Kubelet are both not migrated
2. A/D Controller restarted and migrated (flags flipped)
3. A/D Controller continues to use in-tree code for this node b/c node
   annotation doesn't exist
4. Node drained and made unschedulable. All volumes unmounted/detached with in-tree code
5. Kubelet restarted and migrated (flags flipped)
6. Kubelet annotates node to tell A/D controller this node has been migrated
7. Kubelet is made schedulable.
8. Both A/D Controller & Kubelet Migrated, node is in "fresh" state so all new volumes lifecycle is CSI

Reply:
good suggestion, done